File: iOS/Objective C/IronSourceDemoApp/ViewController.m
//
//  ViewController.m
//  IronSourceDemoApp
//
//  Copyright © 2023 IronSource. All rights reserved.
//

#import "ViewController.h"
#import "RewardedVideoLevelPlayCallbacksHandler.h"
#import "InterstitialLevelPlayCallbacksHandler.h"
#import "BannerLevelPlayCallbacksHandler.h"
#import <IronSource/IronSource.h>

#define USERID @"demoapp"
#define APPKEY @"8545d445"

@interface ViewController () <RewardedVideoLevelPlayCallbacksWrapper, InterstitialLevelPlayCallbacksWrapper, BannerLevelPlayCallbacksWrapper, ISImpressionDataDelegate>

@property (weak, nonatomic) IBOutlet UIButton *showRVButton;
@property (weak, nonatomic) IBOutlet UIButton *loadISButton;
@property (weak, nonatomic) IBOutlet UIButton *showISButton;
@property (weak, nonatomic) IBOutlet UILabel  *versionLabel;

@property (nonatomic, strong) RewardedVideoLevelPlayCallbacksHandler *rewardedVideoDelegate;
@property (nonatomic, strong) InterstitialLevelPlayCallbacksHandler *interstitialDelegate;
@property (nonatomic, strong) BannerLevelPlayCallbacksHandler *bannerDelegate;

@property (nonatomic, strong) ISPlacementInfo   *rvPlacementInfo;
@property (nonatomic, strong) ISBannerView      *bannerView;

@end

@implementation ViewController

#pragma mark -
#pragma mark Lifecycle Methods

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //The integrationHelper is used to validate the integration. Remove the integrationHelper before going live!
    [ISIntegrationHelper validateIntegration];
    
    // UI setup
    self.versionLabel.text = [NSString stringWithFormat:@"sdk version %@", [IronSource sdkVersion]];
    
    for (UIButton *button in @[self.showISButton, self.showRVButton, self.loadISButton]) {
        button.layer.cornerRadius = 17.0f;
        button.layer.masksToBounds = YES;
        button.layer.borderWidth = 3.5f;
        button.layer.borderColor = [[UIColor grayColor] CGColor];
    }
    
    [ISSupersonicAdsConfiguration configurations].useClientSideCallbacks = @(YES);
    
    // Before initializing any of our products (Rewarded video, Interstitial or Banner) you must set
    // their delegates. Take a look at each of these delegates method and you will see that they each implement a product
    // protocol. This is our way of letting you know what's going on, and if you don't set the delegates
    // we will not be able to communicate with you.
    // We're passing 'self' to our delegates because we want
    // to be able to enable/disable buttons to match ad availability.
    
    self.rewardedVideoDelegate = [[RewardedVideoLevelPlayCallbacksHandler alloc] initWithDelegate:self];
    self.interstitialDelegate = [[InterstitialLevelPlayCallbacksHandler alloc] initWithDelegate:self];
    self.bannerDelegate = [[BannerLevelPlayCallbacksHandler alloc] initWithDelegate:self];

    [IronSource setLevelPlayRewardedVideoDelegate:self.rewardedVideoDelegate];
    [IronSource setLevelPlayInterstitialDelegate:self.interstitialDelegate];
    [IronSource setLevelPlayBannerDelegate:self.bannerDelegate];
    [IronSource addImpressionDataDelegate:self];

    NSString *userId = [IronSource advertiserId];
    
    if ([userId length] == 0) {
        //If we couldn't get the advertiser id, we will use a default one.
        userId = USERID;
    }
    
    [IronSource setUserId:userId];
    
    // After setting the delegates you can go ahead and initialize the SDK.
    [IronSource initWithAppKey:APPKEY];
    
    // To initialize specific ad units:
    // [IronSource initWithAppKey:APPKEY adUnits:@[IS_REWARDED_VIDEO, IS_INTERSTITIAL, IS_BANNER]];
    
    // Scroll down the file to find out what happens when you click a button...
    
    /* 
     * Banner integration
     * To finalize your banner integration, you must integrate at least one of our Mediation adapters that have banner.
     */
    [self loadBanner];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark -
#pragma mark Interface Handling

- (IBAction)showRVButtonTapped:(id)sender {
    // After calling 'setLevelPlayRewardedVideoDelegate' and 'initWithAppKey:'
    // you are ready to present an ad. You can supply a placement
    // by calling 'showRewardedVideoWithViewController:placement:', or you can simply
    // call 'showRewardedVideoWithViewController'. 
    // In this case the SDK will use the default placement one created for you.
    [IronSource showRewardedVideoWithViewController:self];
}

- (IBAction)loadISButtonTapped:(id)sender {
    // This will load the Interstitial.
    [IronSource loadInterstitial];
}

- (IBAction)showISButtonTapped:(id)sender {
    // This will present the Interstitial. Unlike Rewarded
    // Videos there are no placements.
    [IronSource showInterstitialWithViewController:self];
}

- (void)loadBanner {
    // We call destroy banner before loading a new banner
    if (self.bannerView) {
        [self destroyBanner];
    }
    
    // This will load the Banner. You can supply a placement
    // by calling 'loadBannerWithViewController:size:placement', or you can simply
    // call 'loadBannerWithViewController:size'. 
    // In this case the SDK will use the default placement one created for you.
    // You can pick any banner size: ISBannerSize_BANNER, ISBannerSize_LARGE, ISBannerSize_RECTANGLE or ISBannerSize_LEADERBOARD
    [IronSource loadBannerWithViewController:self
                                        size:ISBannerSize_BANNER];
}

- (void)destroyBanner {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.bannerView) {
            [IronSource destroyBanner:self.bannerView];
            self.bannerView = nil;
        }
    });
}

#pragma mark - Rewarded Video LevelPlay Callbacks Wrapper Functions

// Once the rewarded video is availalbe you can call the showRewardedVideoWithViewController: API
- (void)rewardedVideoLevelPlayHasAvailableAdWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showRVButton setEnabled:YES];
    });
}

- (void)rewardedVideoLevelPlayHasNoAvailableAd {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showRVButton setEnabled:NO];
    });
}

- (void)rewardedVideoLevelPlayDidOpenWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

// If the rewarded video fails to show check out 'error' for more information and consult
// our knowledge center for help.
- (void)rewardedVideoLevelPlayDidFailToShowWithError:(NSError *)error
                                           andAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)rewardedVideoLevelPlayDidClick:(ISPlacementInfo *)placementInfo
                            withAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)rewardedVideoLevelPlayDidReceiveRewardForPlacement:(ISPlacementInfo *)placementInfo
                                                withAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    self.rvPlacementInfo = placementInfo;
}

- (void)rewardedVideoLevelPlayDidCloseWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    if (self.rvPlacementInfo) {
        UIViewController *rootVC = [[[[UIApplication sharedApplication] delegate] window] rootViewController];
        NSString *message = [NSString stringWithFormat:@"You have been rewarded %d %@", [self.rvPlacementInfo.rewardAmount intValue], self.rvPlacementInfo.rewardName];
        
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Video Reward"
                                                                      message:message
                                                               preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault
                                                   handler:^(UIAlertAction * action) {
        }];
        
        [alert addAction:okAction];
        [rootVC presentViewController:alert animated:NO completion:nil];
        
        self.rvPlacementInfo = nil;
    }
}

#pragma mark - Interstitial LevelPlay Callbacks Wrapper Functions

- (void)interstitialLevelPlayDidLoadWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showISButton setEnabled:YES];
    });
}

- (void)interstitialLevelPlayDidFailToLoadWithError:(NSError *)error {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showISButton setEnabled:NO];
    });
}

- (void)interstitialLevelPlayDidOpenWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialLevelPlayDidShowWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

// If the interstitial fails to show check out 'error' for more information and consult
// our knowledge center for help.
- (void)interstitialLevelPlayDidFailToShowWithError:(NSError *)error
                                          andAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialLevelPlayDidClickWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)interstitialLevelPlayDidCloseWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

#pragma mark - Banner LevelPlay Callbacks Wrapper Functions

- (void)bannerLevelPlayDidLoad:(ISBannerView *)bannerView
                    withAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.bannerView = bannerView;
        CGFloat y = self.view.frame.size.height - (bannerView.frame.size.height / 2.0);
        
        if (@available(ios 11.0, *)) {
             y -= self.view.safeAreaInsets.bottom;
        }
        
        bannerView.center = CGPointMake(self.view.frame.size.width / 2.0, y);
        [self.view addSubview:bannerView];
    });
}

- (void)bannerLevelPlayDidFailToLoadWithError:(NSError *)error {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)bannerLevelPlayDidClickWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)bannerLevelPlayDidLeaveApplicationWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)bannerLevelPlayDidPresentScreenWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

- (void)bannerLevelPlayDidDismissScreenWithAdInfo:(ISAdInfo *)adInfo {
    NSLog(@"%s", __PRETTY_FUNCTION__);
}

#pragma mark - Impression Data Delegate Functions

- (void)impressionDataDidSucceed:(ISImpressionData *)impressionData {
    NSLog(@"impressionData %@", impressionData);
}

@end

File: iOS/Swift/IronSourceSwiftDemoApp/ViewController.swift
//
//  ViewController.swift
//  IronSourceSwiftDemoApp
//
//  Copyright © 2023 IronSource. All rights reserved.
//

import UIKit
import Foundation
import ObjectiveC.runtime
import IronSource

let defaultUserId = "demoapp"
let appKey = "8545d445"

class ViewController: UIViewController, RewardedVideoLevelPlayCallbacksWrapper, InterstitialLevelPlayCallbacksWrapper, BannerLevelPlayCallbacksWrapper, ISImpressionDataDelegate {
    
    //MARK: IBOutlets
    
    @IBOutlet weak var showRewardedVideoButton: UIButton!
    
    @IBOutlet weak var loadInterstitialButton: UIButton!
    @IBOutlet weak var showInterstitialButton: UIButton!
    
    @IBOutlet weak var loadBannerButton: UIButton!
    @IBOutlet weak var destroyBannerButton: UIButton!
    
    @IBOutlet weak var versionLabel: UILabel!
    
    var rewardedVideoDelegate: RewardedVideoLevelPlayCallbacksHandler! = nil
    var interstitialDelegate: InterstitialLevelPlayCallbacksHandler! = nil
    var bannerDelegate: BannerLevelPlayCallbacksHandler! = nil
    
    var bannerView: ISBannerView! = nil
    
    //MARK: ViewLifecycle Functions
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view, typically from a nib.
        self.versionLabel.text =  String(format: "%@%@", "sdk version: ", IronSource.sdkVersion());
        self.setupIronSourceSdk()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //MARK: Private Functions
    
    func setupIronSourceSdk() {
        //The integrationHelper is used to validate the integration. Remove the integrationHelper before going live!
        ISIntegrationHelper.validateIntegration()
        
        // Before initializing any of our products (Rewarded video, Interstitial or Banner) you must set
        // their delegates. Take a look at each of these delegates method and you will see that they each implement a product
        // protocol. This is our way of letting you know what's going on, and if you don't set the delegates
        // we will not be able to communicate with you.
        // We're passing 'self' to our delegates because we want
        // to be able to enable/disable buttons to match ad availability.
        
        rewardedVideoDelegate = .init(delegate: self)
        interstitialDelegate = .init(delegate: self)
        bannerDelegate = .init(delegate: self)
        
        IronSource.setLevelPlayRewardedVideoDelegate(rewardedVideoDelegate)
        IronSource.setLevelPlayInterstitialDelegate(interstitialDelegate)
        IronSource.setLevelPlayBannerDelegate(bannerDelegate)
        IronSource.add(self)
        
        var userId = IronSource.advertiserId()
        
        if userId.count == 0 {
            //If we couldn't get the advertiser id, we will use a default one.
            userId = defaultUserId;
        }
        
        IronSource.setUserId(userId)
        
        // After setting the delegates you can go ahead and initialize the SDK.
        IronSource.initWithAppKey(appKey)
        
        // To initialize specific ad units:
        // IronSource.initWithAppKey(appKey, adUnits:[IS_REWARDED_VIDEO, IS_INTERSTITIAL, IS_BANNER])
        
        // Scroll down the file to find out what happens when you click a button...
    }
    
    func logFunctionName(string: String = #function) {
        print("IronSource Swift Demo App: " + string)
    }
    
    
    func destroyBanner() {
        if bannerView != nil {
            IronSource.destroyBanner(bannerView)
        }
        
        destroyBannerButton.isEnabled = false
    }
    
    //MARK: IBOutlets Actions
    
    @IBAction func showRewardedVideoButtonAction(_ sender: Any) {
        // After calling 'setLevelPlayRewardedVideoDelegate' and 'initWithAppKey'
        // you are ready to present an ad. You can supply a placement
        // by calling 'showRewardedVideo(with: viewController, placement: placement)', or you can simply
        // call 'showRewardedVideo(with:viewController)'.
        // In this case the SDK will use the default placement one created for you.
        IronSource.showRewardedVideo(with: self)
    }
    
    @IBAction func loadInterstitialButtonAction(_ sender: Any) {
        // This will load the Interstitial.
        IronSource.loadInterstitial()
    }

    @IBAction func showInterstitialButtonAction(_ sender: Any) {
        // This will present the Interstitial. Unlike Rewarded
        // Videos there are no placements.
        IronSource.showInterstitial(with: self)
    }
    
    @IBAction func loadBannerButtonAction(_ sender: Any) {
        // We call destroy banner before loading a new banner
        if bannerView != nil {
            destroyBanner()
        }
        
        let bannerSize: ISBannerSize = ISBannerSize(description:"BANNER", width:320, height:50)
        
        // This will load the Banner. You can supply a placement
        // by calling 'loadBanner(with: viewControler, size: bannerSize, placement: placement)', or you can simply
        // call 'loadBanner(with: self, size: bannerSize)'.
        // In this case the SDK will use the default placement one created for you.
        IronSource.loadBanner(with: self, size: bannerSize)
    }
    
    @IBAction func destroyBannerButtonAction(_ sender: Any) {
        destroyBanner()
    }
    
    //MARK: Rewarded Video LevelPlay Callbacks Wrapper Functions
    
    func rewardedVideoLevelPlayHasAvailableAd(with adInfo: ISAdInfo!) {
        showRewardedVideoButton.isEnabled = true
        logFunctionName()
    }
    
    func rewardedVideoLevelPlayHasNoAvailableAd() {
        showRewardedVideoButton.isEnabled = false
        logFunctionName()
    }
    
    func rewardedVideoLevelPlayDidOpen(with adInfo: ISAdInfo!) {
        showRewardedVideoButton.isEnabled = false
        logFunctionName()
    }
    
    func rewardedVideoLevelPlayDidFailToShowWithError(_ error: Error!, andAdInfo adInfo: ISAdInfo!) {
        logFunctionName(string: #function+String(describing: error.self))
    }
    
    func rewardedVideoLevelPlayDidClick(_ placementInfo: ISPlacementInfo!, with adInfo: ISAdInfo!) {
        logFunctionName(string: #function+String(describing: placementInfo.self))
    }
    
    func rewardedVideoLevelPlayDidReceiveRewardForPlacement(_ placementInfo: ISPlacementInfo!, with adInfo: ISAdInfo!) {
        logFunctionName(string: #function+String(describing: placementInfo.self))
    }
    
    func rewardedVideoLevelPlayDidClose(with adInfo: ISAdInfo!) {
        logFunctionName()
    }
    
    
    //MARK: Interstitial LevelPlay Callbacks Wrapper Functions
    
    func interstitialLevelPlayDidLoad(with adInfo: ISAdInfo!) {
        showInterstitialButton.isEnabled = true
        logFunctionName()
    }
    
    func interstitialLevelPlayDidFailToLoadWithError(_ error: Error!) {
        showInterstitialButton.isEnabled = false
        logFunctionName(string: String(describing: error.self))
    }
    
    func interstitialLevelPlayDidOpen(with adInfo: ISAdInfo!) {
        showInterstitialButton.isEnabled = false
        logFunctionName()
    }
    
    func interstitialLevelPlayDidShow(with adInfo: ISAdInfo!) {
        logFunctionName()
    }
    
    func interstitialLevelPlayDidFailToShowWithError(_ error: Error!, andAdInfo adInfo: ISAdInfo!) {
        logFunctionName(string: String(describing: error.self))
    }
    
    func interstitialLevelPlayDidClick(with adInfo: ISAdInfo!) {
        logFunctionName()
    }
        
    func interstitialLevelPlayDidClose(with adInfo: ISAdInfo!) {
        logFunctionName()
    }


    //MARK: Banner LevelPlay Callbacks Wrapper Functions
    
    func bannerLevelPlayDidLoad(_ bannerView: ISBannerView!, andAdInfo adInfo: ISAdInfo!) {
        self.bannerView = bannerView
        
        var xPosition = (view.frame.size.width - bannerView.frame.size.width) / 2.0
        var yPosition = view.frame.size.height - bannerView.frame.size.height
        var bannerWidth = bannerView.frame.size.width
        var bannerHeight = 0.0
        
        if #available(iOS 11.0, *) {
            bannerHeight = bannerView.frame.size.height - self.view.safeAreaInsets.bottom * 2.5
        } else {
            bannerHeight = bannerView.frame.size.height * 2.5
        }
        
        bannerView.frame = CGRect(x: xPosition, y: yPosition, width: bannerWidth, height: bannerHeight)
        view.addSubview(bannerView)
        destroyBannerButton.isEnabled = true
        
        logFunctionName()
    }
    
    func bannerLevelPlayDidFailToLoadWithError(_ error: Error!) {
        logFunctionName(string: #function+String(describing: Error.self))
    }
    
    func bannerLevelPlayDidClick(with adInfo: ISAdInfo!) {
        logFunctionName()
    }
    
    func bannerLevelPlayDidLeaveApplication(with adInfo: ISAdInfo!) {
        logFunctionName()
    }
    
    func bannerLevelPlayDidPresentScreen(with adInfo: ISAdInfo!) {
        logFunctionName()
    }
    
    func bannerLevelPlayDidDismissScreen(with adInfo: ISAdInfo!) {
        logFunctionName()
    }


    //MARK: ISImpressionData Functions

    func impressionDataDidSucceed(_ impressionData: ISImpressionData!) {
        logFunctionName(string: #function+String(describing: impressionData))
    }
}
